tests/unit/test_task_runner.py

```diff
@@ -0,0 +1,167 @@
+import asyncio
+
+import pytest
+import pytest_asyncio
+from pytest import LogCaptureFixture
+from pytest_mock import MockerFixture
+
+from merino.utils.task_runner import gather
+from tests.types import FilterCaplogFixture
+
+# The duration of the slow coroutine (500 ms).
+SLOW_COROUTINE_DURATION = 0.5
+
+
+@pytest_asyncio.fixture(name="normal_task")
+async def fixture_normal_task() -> asyncio.Task:
+    async def normal_op() -> bool:
+        return True
+
+    return asyncio.create_task(normal_op(), name="normal-task")
+
+
+@pytest_asyncio.fixture(name="timedout_task")
+async def fixture_timedout_task() -> asyncio.Task:
+    async def some_slow_op() -> bool:
+        await asyncio.sleep(SLOW_COROUTINE_DURATION)
+        return True
+
+    return asyncio.create_task(some_slow_op(), name="timedout-task")
+
+
+@pytest_asyncio.fixture(name="raised_task")
+async def fixture_raised_task() -> asyncio.Task:
+    async def will_raise() -> None:
+        raise RuntimeError("error")
+
+    return asyncio.create_task(will_raise(), name="raised-task")
+
+
+@pytest.mark.asyncio
+async def test_gather_without_tasks() -> None:
+    """Test gather with an empty task list"""
+    done_tasks, timedout_tasks = await gather([])
+
+    assert done_tasks == []
+    assert timedout_tasks == []
+
+
+@pytest.mark.asyncio
+async def test_gather_tasks_with_timeout(
+    normal_task,
+    timedout_task,
+    raised_task,
+    mocker: MockerFixture,
+    caplog: LogCaptureFixture,
+    filter_caplog: FilterCaplogFixture,
+) -> None:
+    """Test gather with timed out tasks"""
+    stub = mocker.stub(name="timeout_callback")
+
+    done_tasks, timedout_tasks = await gather(
+        [normal_task, timedout_task, raised_task],
+        timeout=SLOW_COROUTINE_DURATION / 2,
+        timeout_cb=stub,
+    )
+
+    assert len(done_tasks) == 2
+    for task in done_tasks:
+        match task.get_name():
+            case "normal-task":
+                assert await task
+            case "raised-task":
+                with pytest.raises(RuntimeError) as exc_info:
+                    await task
+                assert str(exc_info.value) == "error"
```

Comment:
I worry long term using a loop and decision logic like this in test, because it doesn't detect changes, like if a task name changes. Something like the following seems more secure:

```suggestion
    normal_task = next(task for task in done_tasks if task.get_name() == "normal-task")
    assert await normal_task

    raised_task = next(task for task in done_tasks if task.get_name() == "raised-task")
    with pytest.raises(RuntimeError) as exc_info:
        await raised_task
    assert str(exc_info.value) == "error"
```

Reply:
Good catch. Would this do it as well?

```suggestion
    for task in done_tasks:
        match task.get_name():
            case "normal-task":
                assert await task
            case "raised-task":
                with pytest.raises(RuntimeError) as exc_info:
                    await task
                assert str(exc_info.value) == "error"
            case _:
                raise AssertionError(f"Unexpected task: {task.get_name()}")
```

Reply:
🤔 Yes. I think that works.